reconcile/vault_replication.py

```diff
@@ -0,0 +1,225 @@
+    VaultReplicationV1,
+    VaultInstanceV1,
+)
+from reconcile.gql_definitions.vault_policies.vault_policies import (
+    VaultPolicyV1,
+)
+
+from reconcile.utils import gql
+
+from typing import List, cast, Optional
+
+QONTRACT_INTEGRATION = "vault-replication"
+
+
+def deep_copy_versions(
+    dry_run: bool,
+    source_vault,
+    dest_vault,
+    current_dest_version,
+    current_source_version,
+    path,
+):
+    for version in range(current_dest_version + 1, current_source_version + 1):
+        secret_dict = {"path": path, "version": version}
+        copy_secret(dry_run, source_vault, dest_vault, path, version, secret_dict)
+
+
+def copy_secret(
+    dry_run: bool,
+    source_vault: _VaultClient,
+    dest_vault: _VaultClient,
+    path: str,
+    version: int,
+    secret_dict: dict,
+) -> None:
+    secret, version = source_vault.read_all_version(secret_dict)
+    write_dict = {"path": path, "data": secret}
+    logging.info(["replicate_vault_secret", version, path])
+    if not dry_run:
+        dest_vault.write(write_dict)
+
+
+def copy_vault_secret(
+    dry_run: bool, source_vault: _VaultClient, dest_vault: _VaultClient, path: str
+) -> None:
+
+    secret_dict = {"path": path, "version": "LATEST"}
+    _, version = source_vault.read_all_version(secret_dict)
+
+    try:
+        _, dest_version = dest_vault.read_all_version(secret_dict)
+        if dest_version is None and version is None:
+            # Secret is a v1 and does not return version
+            copy_secret(dry_run, source_vault, dest_vault, path, 1, secret_dict)
+        elif dest_version < version:
+            deep_copy_versions(
+                dry_run, source_vault, dest_vault, dest_version, version, path
+            )
+        else:
+            logging.info(["replicate_vault_secret", dest_version, version, path])
+    except SecretNotFound:
+        logging.info(["replicate_vault_secret", "Secret not found", path])
+        deep_copy_versions(dry_run, source_vault, dest_vault, 0, version, path)
+
+
+def check_copy_secret_list(
+    dry_run: bool,
+    source_vault: _VaultClient,
+    dest_vault: _VaultClient,
+    path_list: List[str],
+    policy_paths: Optional[List[str]],
+) -> None:
+
+    invalid_paths = []
+    if policy_paths is not None:
+        invalid_paths = list_invalid_paths(path_list, policy_paths)
+    if invalid_paths:
+        logging.error(["replicate_vault_secret", "Invalid paths", invalid_paths])
+    else:
+        for path in path_list:
+            copy_vault_secret(dry_run, source_vault, dest_vault, path)
+
+
+def list_invalid_paths(path_list: List[str], policy_paths: List[str]) -> List[str]:
+    invalid_paths = []
+    for path in path_list:
+        if not policy_contais_path(path, policy_paths):
+            invalid_paths.append(path)
+
+    return invalid_paths
+
+
+def policy_contais_path(path: str, policy_paths: List[str]) -> bool:
+    if policy_paths is None:
```

Comment:
this is weird. We should not call this function if policy_paths is None

Reply:
Changed the check outside the function instead of checking it for every path